### ISSUE TITLE
Unit and Metal Icons Missing, Black Box Instead

### ISSUE THREAD

**Issue by FireAnt333** [Legion-Expansion/Legion-Expansion#489]
First of all, I'd like to say thank you to all of you who helped develop the PA Legion expansion. I am looking forward to playing it! 

Unfortunately, I've run into an issue. It seems fairly simple, but I haven't seen any other reports of it and I can't find anything that I am doing wrong that could cause it. None of the unit icons (commanders, structures or mobile units) or metal point icons are displayed properly. All I get is a black square where the icon should be. I've tested it on several different maps playing as both Legion and MLA, and the issue is the same. The icons never load in, even after attempting to play for 10+ minutes. 

I installed the mod following the instructions on the Uber forum thread and Exodus E-sports pages. (https://forums.uberent.com/threads/rel-legion-expansion-public-beta.71680/ and http://exodusesports.com/article/legion-expansion-community-faction-mod/) I do not have any other mods installed, and I am running the latest PTE build as Legion requires. 

Is this so easy to fix that it is never reported, or is it something new? Am I doing something wrong and missing something obvious? 

I have attached a screenshot of the issue along with some of the logs from today. I hope that helps! 

Thanks, 
FireAnt333

[PA-2016-05-07_140037.txt](https://github.com/Legion-Expansion/Legion-Expansion/files/253676/PA-2016-05-07_140037.txt)
[PA-2016-05-07_151026.txt](https://github.com/Legion-Expansion/Legion-Expansion/files/253678/PA-2016-05-07_151026.txt)
[PA-2016-05-07_151944.txt](https://github.com/Legion-Expansion/Legion-Expansion/files/253677/PA-2016-05-07_151944.txt)
[PA-2016-05-07_155739.txt](https://github.com/Legion-Expansion/Legion-Expansion/files/253675/PA-2016-05-07_155739.txt)
![pa_legion_icons_missing](https://cloud.githubusercontent.com/assets/19244384/15095191/67dd40ea-1470-11e6-8ec2-60411c0e4661.PNG)

**Comment by Stuart1998** [comment]
This issue is seen if one of the following is the case:

1. You're on Linux.
2. You're on an Intel integrated graphics card.
3. You're using software UI.

Mikeyh or Wondible need to fix it.

**Comment by FireAnt333** [comment]
Thanks for the timely response! In my case I'm pretty sure it's the graphics card. Good luck working out the bugs!

**Comment by JustinLove** [comment]
I've made a test

https://github.com/JustinLove/icon_test/archive/master.zip

Install into C:\Users\<username>\AppData\Local\Uber Entertainment\Planetary Annihilation\client_mods\

Make sure unzipped paths work out so you have ...\client_mods\icon_test\modinfo.json  You can enable in community mods.

Set the minimum number icons in game settings. Changing the value in-game should work. Use without legion, because if it's the number of icons, it will be something less. Standard PA has 121 icons and legion has 234; if 234 triggers its the number of icons, and you can try to narrow down exactly where it cuts out.

On OSX with Intel HD5000 I can get 315 icons before they go black. I can go over, lower the amount, and the icons will return.

**Comment by Quitch** [comment]
No response received.

**Comment by FireAnt333** [comment]
My apologies, over the last several days I have been unable to play or try the test that was made. I should have let you know I was unavailable.

**Comment by FireAnt333** [comment]
After testing, here is my report: 

I installed the test file as you instructed, into C:\Users<username>\AppData\Local\Uber Entertainment\Planetary Annihilation\client_mods\. There was a bit of an issue with this, because I couldn't find a "client_mods" folder within C:\Users<username>\AppData\Local\Uber Entertainment\Planetary Annihilation. I had to create that folder myself, then put the files you provided into that new folder. (Let me know if that invalidates anything in the testing.) When I started up the game with the PTE build, there were two yellow boxes that came up on the main menu that I hadn't seen before, but they didn't seem to cause any problems. The first one said "PLEASE CHANGE THE CONTENTS OF VERSION.TXT IN YOUR PTE INSTALL TO 89755". The second one said "TO USE THE PREVIEW VERSION OF COMMUNITY MODS PLEASE UNINSTALL ALL OF YOUR MODS IN PAMM THEN RESTART PA". I was able to find the "Icon Test" client mod in the Community Mods section, and enabled it. 

I started a game (PA: Titans without Legion) and everything worked fine. I'm not sure what you meant about the icons, whether it was the total number of icons displayed, or the number of different icons installed, but I got around 500 icons to display at once, including about 400 Dox plus metal points, metal extractors and factories. 

![pa_icon_test](https://cloud.githubusercontent.com/assets/19244384/15407333/299f2d04-1dbf-11e6-8668-d184b6726e6b.PNG)

I'm sorry it took me so long to finally do this! Should I test it with the Legion expansion enabled as well? 

FireAnt333

**Comment by JustinLove** [comment]
Thanks for testing, and sorry for my poor instructions.

I added an item in settings, Gameplay, for the number of distinct icons types (dox vs ant etc) Legion roughly doubles how many different types of icons the game needs to display, and the only time I recall seeing black boxes myself is when I pushed the number of icons to an even higher level. Legion only gives us one data point for number of icons. You'll have to manipulate that extra item in settings; nothing will change in-game, except if you find an icon limit and eveything goes black.

**Comment by FireAnt333** [comment]
That's okay, I should have been able to figure it out! I knew from your original instructions that I had to change something in the game settings, but I forgot to even look there! When you said to "change the value in-game," I thought you meant changing the actual number of icons by building more units! I guess I got that wrong. 

After trying it again, this time the right way, this is what I found: 

I set the minimum number of icons to various numbers, and this is what happened: 

100: No change
150: Metal spot icon stretched vertically and displayed slightly below actual metal spot; structure and unit icons displayed in pieces and either above, below or with pieces both above and below the unit or structure. 
200: Black squares identical to those in the original problem.  

When I tested it again with a different player color to get a better view of the icons (orange instead of navy blue) and got some screenshots. 

100: 
![pa_icon_test_2](https://cloud.githubusercontent.com/assets/19244384/15413286/95fa4d84-1de3-11e6-848f-744d1eb61ad7.PNG)

150: 
![pa_icon_test_3](https://cloud.githubusercontent.com/assets/19244384/15413288/9b8659dc-1de3-11e6-822b-6fc55fbe5423.PNG)

200: 
![pa_icon_test_4](https://cloud.githubusercontent.com/assets/19244384/15413294/a7495490-1de3-11e6-9736-cd1dd2865099.PNG)

Narrowing it down further, 121 appeared normal: 
![pa_icon_test_5](https://cloud.githubusercontent.com/assets/19244384/15413339/07fb01d0-1de4-11e6-85ba-ff5b94aef29a.PNG)

At 122, as you said, just one above the normal icon count in PA, the problems start up. Some seem nearly unaffected, (the dox, bot factory and energy plant) while some are a little more off (metal extractor) and some are more problematic (commander). 
![pa_icon_test_6](https://cloud.githubusercontent.com/assets/19244384/15413389/9670d764-1de4-11e6-8802-9d0496826bc5.PNG)

I played around with it more and got some strange behavior. I kicked it up to 150, and the icons went back to normal. I adjusted to 151, and they stayed normal. At 150 and 151 they would show the expected misalignment and pieces of the wrong icons, but only for a split second before returning to the normal, proper display. I increased the number to 160, and they stayed normal for a couple of seconds, then they all switched to the black squares. 

I tried to find the exact number where it cut out to black, and got some more weird behavior. 

At 155, I got the expected distortion: 
![pa_icon_test_7](https://cloud.githubusercontent.com/assets/19244384/15413563/5fa652e8-1de6-11e6-9b8b-55ab2d004e20.PNG)

But at 156, it did the split-second glitch and returned to normal: 
![pa_icon_test_8](https://cloud.githubusercontent.com/assets/19244384/15413547/4277e858-1de6-11e6-9f1e-2fbb55be0a91.PNG)

I reset it to 155 and got the same result that 155 previously gave, then tried 156 again and it did what I originally expected: 
![pa_icon_test_9](https://cloud.githubusercontent.com/assets/19244384/15413585/a0143c28-1de6-11e6-873c-e6f8b95bd746.PNG)

It progressed to 158, slightly changing each time, until it finally cut out to black at 159: 
![pa_icon_test_10](https://cloud.githubusercontent.com/assets/19244384/15413611/e77596e8-1de6-11e6-9f35-2201bcc141a1.PNG)

When I stepped back down to 158, the black squares stayed rather than going back to misaligned icons. Stepping down to 157 caused icons to become totally normal again as soon as I exited the settings menu, then they displayed the expected misalignment for a split second, then went back to normal again. For some reason it doesn't appear to be exactly precise, but misalignment starts around 120 and cut-out to black starts around 160. 

I hope this helps! Sorry it took so long to get the right testing information back to you! 
FireAnt333

**Comment by JustinLove** [comment]
Okay, for the black boxes it's looking like my texture size theory is holding up, and the number is system specific.

We'd need to reintroduce an icon manager for aliasing. Cutting off the icons and leaving some as dots would be much simpler.  Anything will need a setting for systems that can't handle all the legion icons.

The misalignment might be a timing issue, although I thought the updates took care of that. I haven't seen the stretching before, but it might be related. I might be able to make a version of the test which allows adjusting delays.

**Comment by mdsitton** [comment]
I've been having this bug on one of the systems I play on. I don't know that much about modding PA, but I may be able to help out trying to diagnose this if it is OpenGL related. So far it looks like the texture atlas that the game has the icons stacked vertically, with the x texture coords the same for all icons.

It might be worth checking out the max texture size Which you might be able to check with webgl and calculate the max number of icons. The opengl value you would check to get the max is GL_MAX_TEXTURE_SIZE (idk about webgl).

I'm actually not sure how the drivers interpret max texture size, lets say for some reason its the absolute max for the x or y number of pixels. Then PA with a large number of icons could well be reaching past this.

If we have 253 icons (counted from the number of icons in the vanilla and legion folders) multiply that by a height of 52px that would give us an atlas height of 13156px. Lets say we use the value @FireAnt333 got before the black icons: 157. That gives us a texture height of 8164, mighty close to 8192. Less than an icon from the max.

So basically to properly play legion you need a max texture hight of at least 16384, this gives a limit of 315 icons.

This could be fixed if PA's code were changed to use a texture atlas that would put icons in the x an y axis in the atlas...

**Comment by mdsitton** [comment]
Also later i'm going to write a test in C++/OpenGL to test my theory...

**Comment by FireAnt333** [comment]
I had a similar, though more primitive theory similar to that, but because of my limited experience in coding, I figured I would leave it to the experts. 

I would like to ask, though, are there any reasons why a vertically-stacked texture atlas would be used instead of a grid-based one, other than simplicity? Would it be possible to change to a grid-based texture atlas within the Legion mod?

**Comment by mdsitton** [comment]
Again I dont really know too much about modding PA and whats possible. But pretty sure this is an engine issue.

**Comment by mdsitton** [comment]
Alright i've started working on a proof of concept thing but its mainly just setup code for sdl2 and glew atm. Tomarrow, i'll start worklng on some rendering code. https://github.com/mdsitton/atlas-test

I can also confirm that the computer that i have the black icon issue on has a max texture size of 8192.

**Comment by JustinLove** [comment]
I've gotten a few instances of the misasligned stretched icons, but it won't reproduce reliably enough for me to say with any confidence that it's a particular number of icons, or that something I've done has solved the issue. If you can get consistent reproductions at certain icon numbers, I've got an updated version of the icon test which allows you to change the delay between updating the atlas and telling the engine to update it's index mapping. My leading theory is that we are catching an odd moment in the atlas update process where things haven't settled down yet.

https://github.com/JustinLove/icon_test

**Comment by JustinLove** [comment]
Just for confirmation: my laptop's max texture size is 16384x16384, as expected.

**Comment by JustinLove** [comment]
In theory it should be detectable from PA, though I haven't yet determined how much code it will take; try this in a debugger console:

window.location = 'https://www.browserleaks.com/webgl'

**Comment by JustinLove** [comment]
I've updated the test again. In settings it reports MAX_TEXTURE_SIZE (which came out 4096 on my system, so it may be a limit in this version of coherent) and MAX_RENDERBUFFER_SIZE, which is a metric that matched the expected value. There is also a convenience link to the webgl report I used above, in case neither of those numbers match the expected value and you need to look for a better one.

**Comment by FireAnt333** [comment]
I wasn't sure if you need any additional information, so I decided to include everything that link gave me, and everything I saw show up in the in-game settings. 

------------------------------
WebGL Support Detection
Does Your Browser Supports WebGL?	✔ Yes
Is WebGL Enabled in Your Browser?	✔ Yes
What about WebGL 2.0 Support?	× No
Browser User-Agent:	Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
WebGL Context Info
Supported Context Name(s)	webgl, experimental-webgl
GL Version	WebGL 1.0 (OpenGL ES 2.0 Chromium)
Shading Language Version	WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor	WebKit
Renderer	WebKit WebGL
Debug Renderer Info
Unmasked Vendor	! Google Inc.
Unmasked Renderer	! ANGLE (Intel(R) HD Graphics 4000 Direct3D9Ex vs_3_0 ps_3_0)
Vertex Shader
Max Vertex Attributes	16
Max Vertex Uniform Vectors	253
Max Vertex Texture Image Units	4
Max Varying Vectors	9
Best Float Precision	[-2127, 2127] (23)
Rasterizer
Aliased Line Width Range	[1, 1]
Aliased Point Size Range	[1, 256]
Fragment Shader
Max Fragment Uniform Vectors	221
Max Texture Image Units	16
Float/Int Precision:	highp/highp
Best Float Precision	[-2127, 2127] (23)
Framebuffer
Max Color Buffers:	1
RGBA Bits	[8, 8, 8, 8]
Depth / Stencil Bits:	[24, 8]
Max Render Buffer Size	8192
Max Viewport Dimensions	[8192, 8192]
Textures
Max Texture Size	8192
Max Cube Map Texture Size	8192
Max Combined Texture Image Units	20
Max Anisotropy	16
Miscellaneous :
Antialiasing	Available
ANGLE	Yes, Direct3D 9
Major Performance Caveat	No
Supported WebGL Extensions	ANGLE_instanced_arrays
EXT_blend_minmax
EXT_frag_depth
EXT_shader_texture_lod
EXT_texture_filter_anisotropic
WEBKIT_EXT_texture_filter_anisotropic
OES_element_index_uint
OES_standard_derivatives
OES_texture_float
OES_texture_float_linear
OES_texture_half_float
OES_texture_half_float_linear
OES_vertex_array_object
WEBGL_compressed_texture_s3tc
WEBKIT_WEBGL_compressed_texture_s3tc
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_depth_texture
WEBKIT_WEBGL_depth_texture
WEBGL_lose_context
WEBKIT_WEBGL_lose_context
------------------------------

![pa_icon_test_settings](https://cloud.githubusercontent.com/assets/19244384/15594384/fd61c25a-2369-11e6-9c45-276933d421a2.PNG)

**Comment by JustinLove** [comment]
> Would you like me to look for any behavior changes in-game with the new delay between the atlas update and index mapping update?

If you can get the offset/stretched icons repeatedly enough that you could tell whether something had fixed it, it you can try smaller or larger delays and see it helps.

Though if noone has noticed this effect in Legion, it may just be an artifact of the test program.

**Comment by mdsitton** [comment]
Well when I finish writing my test application over here we can confirm atleast if the black icon issue is an opengl issue or not...
https://github.com/mdsitton/atlas-test

**Comment by aldrik** [comment]
**Possible workaround**: Try opening the debugger and running the following commands in the Icon_Atlas console (for half size icons).
```
$('head').append('<style>img { width: 26px !important; height: 26px !important; }</style>');
 engine.call('handle_icon_list', model.strategicIcons(), 26);
```

**Comment by Stuart1998** [comment]
That workaround works.

**Comment by JustinLove** [comment]
I got a two dimensional atlas to work, but it didn't increase the number of available icons, and at some multiples decreased it, always breaking when an additional row was added. I suspect that the real limit is total texture size. For instance if 52 width was being rounded up to 64, my system would be addressing around 1,048,576 pixels. if it was that simple though I should able to get 361 icons into a 1024x1024 texture, but it caps out at 305. So there is no get-out-of-jail from horizontal expansion.

**Comment by Stuart1998** [comment]
35 is the maximum allowed icon size.

**Comment by Spirittracker21** [comment]
I am also having this problem i tried to instal the file to increase the icon limit but did not work

**Comment by mdsitton** [comment]
Just thought i would mention on my system exceding the texture size from the GL_MAX_TEXTURE_SIZE value on any demension gives you GL_INVALID_VALUE from glGetError()

So my inital theory seems to be somewhat correct atleast. I havent quite had the time to finish my test project though as I decided to start benge watching all of the seasons of Game of Thrones for the first time...

**Comment by mdsitton** [comment]
@JustinLove mind if you share how you implemented a 2d atlas?

**Comment by mdsitton** [comment]
So if we can have a 2d atlas. We should be able to use the whole supported texture size 8192x8192 (creation works on my system without error) we would have a max of 24649 52px icons wat above any reasonable amount of unique icons...

**Comment by JustinLove** [comment]
2D does not help, the limit is the same or lower. I think my systems are able to address somewhere in the 850k pixels.There might be dimensions that eek out a extra icon or two, but such small gains aren't worth potential conflicts with icon mods.

Implementation was paired changes to icon_atlas style and particle_icon.vs. The game is feeding somewhat baked values into the shader, but it's possible to calculate the icon index from them. I may wrap it up to a repo at some point for reference.

**Comment by mdsitton** [comment]
@JustinLove First of all do you even know what I mean by a 2d atlas vs a 1D atlas? I've looked at the icon_atlas javascript style and html files and there is no obvious way to do the conversion I speak of within these file.

Which means whatever your doing is not a true 2d texture atlas. What I have been saying this whole time is that if PA in the engine is where this needs to be done

here is 2 images generated with gimp showing what the difference between a 1d and a 2d texture atlas is if you would view the whole texture at once:
http://imgur.com/a/Xw74u

**Comment by JustinLove** [comment]
Reference screenshot I made while working:

![pax16](https://cloud.githubusercontent.com/assets/54438/15780387/fc5a3cfe-2966-11e6-94f8-0c1278912eb8.png)

Code pulled out into it's own mod. I was working on it with the shaders moved to the PA core on a windows machine to get live shader updates. It's currently in x16, which only gets 304 icons (19 rows) before black boxes.

https://github.com/JustinLove/wide_atlas

**Comment by mdsitton** [comment]
So I traced PA with apitrace.
The function call that fails on my system fails with these arguments.
glTexImage2D(target = GL_TEXTURE_2D, level = 0, internalformat = GL_RGBA, width = 52, height = 12168, border = 0, format = GL_RGBA, type = GL_UNSIGNED_BYTE, pixels = blob(2530944))

Whenevr I figure out how to install mod seperate of the built in mod manager I might test it with your test mod as well.

**Comment by JustinLove** [comment]
How do Legion peeps want to deal with this?

So far (in very limited testing) there appears to be a metic accessible by WebGL which correlates with reduced icons, which could be used to trigger different behavior.  We could:

1. Cut off and have dots - assuming this is undesirable, though still better than black boxes.
2. Reintroduce an icon manager in order to have fallbacks. This is gnarly enough that I would want an independent mod so I can do maintenance without worrying about old versions in several different mods. Or maybe Legion will be the only thing big enough to require it?
3. Reduce icon size. This should be simple enough to build in, though it may have be revisited if several mods start playing with it.

**Comment by Stuart1998** [comment]
Icon size of 35 is pretty small and is undesirable. An icon manager would be good. In the past, ModX caused the black boxes. I've been carefully conserving icons in PO to avoid the issue, but it seems inevitable that it will come up again after I add more stuff so an icon manager would be good.

**Comment by Quitch** [comment]
This was integrated into the release of 0.18.0.
